README
RTMP Dump v2.4.1
(C) 2009 Andrej Stepanchuk
(C) 2009-2011 Howard Chu
(C) 2010 2a665470ced7adb7156fcef47f8199a6371c117b8a79e399a2771e0b36384090
(C) 2011 33ae1ce77301f4b4494faaa5f609f3c48b9dcf82
License: GPLv2
librtmp license: LGPLv2.1
http://rtmpdump.mplayerhq.hu/

To compile type "make" with SYS=<platform name>, e.g.

  $ make SYS=posix

for Linux, Unix, etc. or

  $ make SYS=darwin

for MacOSX or

  $ make SYS=mingw

for Windows.

You can cross-compile for other platforms using the CROSS_COMPILE variable:

  $ make CROSS_COMPILE=arm-none-linux- INC=-I/my/cross/includes

Please read the Makefile to see what other make variables are used.

This code also requires you to have OpenSSL and zlib installed. You may
optionally use GnuTLS or polarssl instead of OpenSSL if desired. You may
also build with just rtmpe support, and no rtmps/https support, by
specifying -DNO_SSL in the XDEF macro, e.g.

  $ make XDEF=-DNO_SSL

or

  $ make CRYPTO=POLARSSL XDEF=-DNO_SSL

You may also turn off all crypto support if desired

  $ make CRYPTO=

A shared library is now built by default, in addition to the static
library. You can also turn it off if desired

  $ make SHARED=

The rtmpdump programs still link to the static library, regardless.

Note that if using OpenSSL, you must have version 0.9.8 or newer.
For Polar SSL you must have version 1.0.0 or newer.

Credit goes to team boxee for the XBMC RTMP code originally used in RTMPDumper.
The current code is based on the XBMC code but rewritten in C by Howard Chu.


SWF Verification
----------------

Note: these instructions for manually generating the SWFVerification
info are provided only for historical documentation. The software can now
generate this info automatically, so it is no longer necessary to
run the commands described here. Just use the -W (--swfVfy) option
to perform automatic SWFVerification.

Download the swf player you want to use for SWFVerification, unzip it using

 $ flasm -x file.swf

It will show the decompressed filesize, use it for --swfsize

Now generate the hash

 $ openssl sha -sha256 -hmac "Genuine Adobe Flash Player 001" file.swf

and use the --swfhash "01234..." option to pass it.

e.g. $ ./rtmpdump --swfhash "123456..." --swfsize 987...


Connect Parameters
------------------

Some servers expect additional custom parameters to be attached to the
RTMP connect request. The "--auth" option handles a specific case, where
a boolean TRUE followed by the given string are added to the request.
Other servers may require completely different parameters, so the new
"--conn" option has been added. This option can be set multiple times
on the command line, adding one parameter each time.

The argument to the option must take the form <type> : <value> where
type can be B for boolean, S for string, N for number, and O for object.
For booleans the value must be 0 or 1. Also, for objects the value must
be 1 to start a new object, or 0 to end the current object.

Examples:
  --conn B:0 --conn S:hello --conn N:3.14159

Named parameters can be specified by prefixing 'N' to the type. Then the
name should come next, and finally the value:
  --conn NB:myflag:1 --conn NS:category:something --conn NN:pi:3.14159

Objects may be added sequentially:
  -C O:1 -C NB:flag:1 -C NS:status:success -C O:0 -C O:1 -C NN:time:12.30 -C O:0
or nested:
  -C O:1 -C NS:code:hello -C NO:extra:1 -C NS:data:stuff -C O:0 -C O:0


Building OpenSSL 0.9.8k
-----------------------
arm:
./Configure -DL_ENDIAN --prefix=`pwd`/armlibs linux-generic32

Then replace gcc, cc, ar, ranlib in Makefile and crypto/Makefile by arm-linux-* variants  and use make && make install_sw

win32:
Try ./Configure mingw --prefix=`pwd`/win32libs -DL_ENDIAN -DOPENSSL_NO_HW
Replace gcc, cc, ... by mingw32-* variants in Makefile and crypto/Makefile
make && make install_sw

OpenSSL cross-compiling can be a difficult beast.

Precompiled OpenSSL binaries for Windows are available on
http://www.slproweb.com/products/Win32OpenSSL.html

If you're just running a pre-built Windows rtmpdump binary, then all you
need is the "Light" installer. If you want to compile rtmpdump yourself,
you'll need the full installer.


Example Servers
---------------
Three different types of servers are also present in this distribution:
 rtmpsrv - a stub server
 rtmpsuck - a transparent proxy
 rtmpgw - an RTMP to HTTP gateway

rtmpsrv - Note that this is very incomplete code, and I haven't yet decided
whether or not to finish it. It is useful for obtaining all the parameters
that a real Flash client would send to an RTMP server, so that they can be
used with rtmpdump. The current version now invokes rtmpdump automatically
after parsing a client request.

rtmpsuck - proxy server. See below...

All you need to do is redirect your Flash clients to the machine running this
server and it will dump out all the connect / play parameters that the Flash
client sent. The simplest way to cause the redirect is by editing /etc/hosts
when you know the hostname of the RTMP server, and point it to localhost while
running rtmpsrv on your machine. (This approach should work on any OS; on
Windows you would edit %SystemRoot%\system32\drivers\etc\hosts.)

On Linux you can also use iptables to redirect all outbound RTMP traffic. You
need to be running as root in order to use the iptables command.

In my original plan I would have the transparent proxy running as a special
user (e.g. user "proxy"), and regular Flash clients running as any other user.
In that case the proxy would make the connection to the real RTMP server. The
iptables rule would look like this:

iptables -t nat -A OUTPUT -p tcp --dport 1935 -m owner \! --uid-owner proxy \
 -j REDIRECT

A rule like the above will be needed to use rtmpsuck. Note that you should
replace "proxy" in the above command with an account that actually exists
on your machine.

Using it in this mode takes advantage of the Linux support for IP redirects;
in particular it uses a special getsockopt() call to retrieve the original
destination address of the connection. That way the proxy can create the
real outbound connection without any other help from the user. The equivalent
functionality may exist on other OSs but needs more investigation.

(Based on reading the BSD ipfw manpage, this rule ought to work on BSD:

ipfw add 40 fwd 127.0.0.1,1935 tcp from any to any 1935 not uid proxy

Some confirmation from any BSD users would be nice.)

(We have a solution for Windows based on a TDI driver; this is known to
work on Win2K and WinXP but is assumed to not work on Vista or Win7 as the
TDI is no longer used on those OS versions. Also, none of the known
solutions are available as freeware.)

The rtmpsuck command has only one option: "-z" to turn on debug logging.
It listens on port 1935 for RTMP sessions, but you can also redirect other
ports to it as needed (read the iptables docs). It first performs an RTMP
handshake with the client, then waits for the client to send a connect
request. It parses and prints the connect parameters, then makes an
outbound connection to the real RTMP server. It performs an RTMP handshake
with that server, forwards the connect request, and from that point on it
just relays packets back and forth between the two endpoints.

It also checks for a few packets that it treats specially: a play packet
from the client will get parsed so that the playpath can be displayed. It
also handles SWF Verification requests from the server, without forwarding
them to the client. (There would be no point, since the response is tied to
each session's handshake.)

Once the play command is processed, all subsequent audio/video data received
from the server will be written to a file, as well as being delivered back
to the client.

The point of all this, instead of just using a sniffer, is that since rtmpsuck
has performed real handshakes with both the client and the server, it can
negotiate whatever encryption keys are needed and so record the unencrypted
data.

rtmpgw - HTTP gateway: this is an HTTP server that accepts requests that
consist of rtmpdump parameters. It then connects to the specified RTMP
server and returns the retrieved data in the HTTP response. The only valid
HTTP request is "GET /" but additional options can be provided in normal
URL-encoded fashion. E.g.
  GET /?r=rtmp:%2f%2fserver%2fmyapp&y=somefile HTTP/1.0

is equivalent the rtmpdump parameters "-r rtmp://server/myapp -y somefile".

Note that only the shortform (single letter) rtmpdump options are supported.


ChangeLog
RTMPDump
Copyright 2008-2009 Andrej Stepanchuk; Distributed under the GPL v2
Copyright 2009-2011 Howard Chu
Copyright 2009 The Flvstreamer Team
http://rtmpdump.mplayerhq.hu/

9 march 2021, v2.4.1
- added support for openssl-1.1

1 June 2015
- fitted for multimedia player movian


20 July 2011
- add NetStream.Authenticate.UsherToken for Justin.tv

11 July 2011, v2.4
- add RTMPE type 9 handshake support

30 June 2010, v2.3
- fix RC4 cleanup for GnuTLS/gcrypt
- declare RTMP_Write buf as const
- cleanup Makefile
- replace all use of bool with int
- add RTMP_Socket() and RTMP_Pause() APIs
- add ping/pong message handling
- add basic shared library support
- fix RTMP_ParseURL extension skipping
- fix bad switch() from -r477 in RTMP_Read
- fix rtmpsrv to always use .flv extension on output
- fix crash on socket failure while RTMP_Read() is reading header
- fix RTMP_ReadPacket signed/unsigned chars for m_nChannel
- license cleanup, use current FSF address
- fix RTMP_SetupURL tcUrl generation
- ignore multiple spaces between URL options
- only send CheckBW request once per session

29 May 2010, v2.2e
- port to Xbox
- add explicit URL scheme for rtmpts
- fix rtmpt clientID NUL-termination
- use BufferEmpty trick in rtmpdump but not by default in librtmp
- add librtmp manpage
- fix RTMP_Read, return 0 on EOF, not -1
- change RTMP_Read to return 1 packet at a time, fix buffer mgmt
- fix request/result queue, fix server compatibility for Publishing
- add RTMP_EnableWrite to enable Publishing

29 April 2010, v2.2d
- add RTMP_Alloc, RTMP_Free APIs
- add optional support for polarssl instead of OpenSSL
- add option to build crypto support without SSL/TLS
- tweak handshake offset checking
- add RTMP set_playlist command
- check for (and fix) broken timestamps in FLV packets
- fix tcUrl and playpath parsing in rtmpsrv and rtmpsuck
- change internal boolean flags to bitmasks

14 April 2010, v2.2c
- internal restructuring, fix #undef CRYPTO builds
- add RTMP_SetupURL, RTMP_SetOpt APIs
- add logging callback

22 March 2010, v2.2b
- fix v2.2a crashes in rtmpsrv/rtmpsuck
- fix v2.2a .swfinfo location on Windows
- fix typo for --auth parameter in manpages
- add FP10 handshake support for rtmpsrv/rtmpsuck
- avoid GNUMake vs BSDMake incompatibilities
- add pkgconfig file for librtmp
- more library cleanup

20 March 2010, v2.2a
- fix C++ compatibility for librtmp
- misc library restructuring
- add client support for tunneling: rtmpt, rtmpte, rtmps
- fix rtmpdump/rtmpgw FLV header dataType
- implement RTMP_Read() and RTMP_Write() to simplify library use
- fix SendPacket timestamps
- add optional support for GnuTLS/Gcrypt instead of OpenSSL
- use $HOMEPATH on Windows instead of $HOME for .swfinfo

4 March 2010, v2.2
- move RTMP code into library librtmp
- relicense RTMP code under LGPL v2.1
- add rtmpdump manpage
- fix AMF_LONG_STRING handling
- more FlashPlayer 10 handshake support
- in rtmpsrv fix Play Start/Stop messages
- rename "streams" program to "rtmpgw"

20 February 2010, v2.1d
- extend .swfinfo file format, add --swfAge rtmpdump parameter
  old file should be replaced or manually updated: copy the
  "date:" line and rename it to "ctim:"

- fix MacOSX builds - just use "make posix" now for all Unix-derived systems
- more explicit error checks in HTTP_get()
- in rtmpsrv spawn rtmpdump automatically
- fix bug in retry/resume of audio-only streams
- other minor misc. fixes

9 January 2010, v2.1c
- cleanup rtmpsrv output
- fix crash in 2.1b hashswf
- fix parseurl to url-decode PlayPath
- fix parseurl to recognize extensions followed by URL params
- fix Makefile, inadvertently dropped 'v' from version string
- in rtmpdump try Reconnect if ToggleStream doesn't work on timeouts
- in rtmpsuck use chunk-based I/O for better latency
- in rtmpsuck support lists of streams
- in rtmpsuck use raw client connect packet to workaround unsupported features
- support arbitrary AMF data appended to connect requests

4 January 2010, v2.1b
- fix url matching in .swfinfo lookup
- fix resume parsing in rtmpdump
- minor code cleanup (CRYPTO dependencies, logging)
- add getStreamLength recognition to rtmpsrv
- add close processing in rtmpsuck

1 January 2010, v2.1a
- fix socket receive timeouts for WIN32
- add streams description to README

29 December 2009, v2.1
- AMF cleanup: bounds checking for all encoders, moved AMF_EncodeNamed* from rtmp.c
- added SecureToken support
- added automatic SWF hash calculation
- added server-side handshake processing
- added rtmpsrv stub server example
- added rtmpsuck proxy server
- tweaks for logging
- renamed more functions to cleanup namespace for library use
- tweaks for server operation: objectEncoding, chunksize changes

16 December 2009, v2.0
- rewrote everything else in C, reorganized to make it usable again as a library
- fixed more portability bugs
- plugged memory leaks

2 December 2009, v1.9a
- fix auth string typo
- handle FCUnsubscribe message
- don't try retry on live streams
- SIGPIPE portability fix
- remove "not supported" comment for RTMPE

13 November 2009, v1.9
- Handle more signals to reduce risk of unresumable/corrupted partially streamed files
- Fixed >2GB file handling

librtmp/hashswf.c
/*
 *  Copyright (C) 2009-2010 Howard Chu
 *
 *  This file is part of librtmp.
 *
 *  librtmp is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as
 *  published by the Free Software Foundation; either version 2.1,
 *  or (at your option) any later version.
 *
 *  librtmp is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with librtmp see the file COPYING.  If not, write to
 *  the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 *  Boston, MA  02110-1301, USA.
 *  http://www.gnu.org/copyleft/lgpl.html
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <time.h>

#include "rtmp_sys.h"
#include "log.h"
#include "http.h"

#include "networking/net.h"

#ifdef CRYPTO
#ifdef USE_POLARSSL
#include <polarssl/sha256.h>
#ifndef SHA256_DIGEST_LENGTH
#define SHA256_DIGEST_LENGTH	32
#endif
#define HMAC_CTX	sha256_context
#define HMAC_setup(ctx, key, len)	sha256_hmac_starts(&ctx, (unsigned char *)key, len, 0)
#define HMAC_crunch(ctx, buf, len)	sha256_hmac_update(&ctx, buf, len)
#define HMAC_finish(ctx, dig, dlen)	dlen = SHA256_DIGEST_LENGTH; sha256_hmac_finish(&ctx, dig)
#define HMAC_close(ctx)
#elif defined(USE_GNUTLS)
#include <nettle/hmac.h>
#ifndef SHA256_DIGEST_LENGTH
#define SHA256_DIGEST_LENGTH	32
#endif
#undef HMAC_CTX
#define HMAC_CTX	struct hmac_sha256_ctx
#define HMAC_setup(ctx, key, len)	hmac_sha256_set_key(&ctx, len, key)
#define HMAC_crunch(ctx, buf, len)	hmac_sha256_update(&ctx, len, buf)
#define HMAC_finish(ctx, dig, dlen)	dlen = SHA256_DIGEST_LENGTH; hmac_sha256_digest(&ctx, SHA256_DIGEST_LENGTH, dig)
#define HMAC_close(ctx)
#else	/* USE_OPENSSL */
#include <openssl/ssl.h>
#include <openssl/sha.h>
#include <openssl/hmac.h>
#include <openssl/rc4.h>
#define HMAC_setup(ctx, key, len)	HMAC_CTX_init(&ctx); HMAC_Init_ex(&ctx, (unsigned char *)key, len, EVP_sha256(), 0)
#define HMAC_crunch(ctx, buf, len)	HMAC_Update(&ctx, (unsigned char *)buf, len)
#define HMAC_finish(ctx, dig, dlen)	HMAC_Final(&ctx, (unsigned char *)dig, &dlen);
#define HMAC_close(ctx)	HMAC_CTX_cleanup(&ctx)
#endif

extern void RTMP_TLS_Init();
extern TLS_CTX RTMP_TLS_ctx;

#include <zlib.h>

#endif /* CRYPTO */

#define	AGENT	"Mozilla/5.0"

HTTPResult
HTTP_get(struct HTTP_ctx *http, const char *url, HTTP_read_callback *cb)
{
  char *host, *path;
  char *p1, *p2;
  char hbuf[256];
  int port = 80;
#ifdef CRYPTO
  int ssl = 0;
#endif
  int hlen, flen = 0;
  int rc, i;
  int len_known;
  HTTPResult ret = HTTPRES_OK;
  struct sockaddr_in sa;
  RTMPSockBuf sb = {0};

  http->status = -1;

  memset(&sa, 0, sizeof(struct sockaddr_in));
  sa.sin_family = AF_INET;

  /* we only handle http here */
  if (strncasecmp(url, "http", 4))
    return HTTPRES_BAD_REQUEST;

  if (url[4] == 's')
    {
#ifdef CRYPTO
      ssl = 1;
      port = 443;
      if (!RTMP_TLS_ctx)
	RTMP_TLS_Init();
#else
      return HTTPRES_BAD_REQUEST;
#endif
    }

  p1 = strchr(url + 4, ':');
  if (!p1 || strncmp(p1, "://", 3))
    return HTTPRES_BAD_REQUEST;

  host = p1 + 3;
  path = strchr(host, '/');
  hlen = path - host;
  strncpy(hbuf, host, hlen);
  hbuf[hlen] = '\0';
  host = hbuf;
  p1 = strrchr(host, ':');
  if (p1)
    {
      *p1++ = '\0';
      port = atoi(p1);
    }


  sb.sb_tcp = tcp_connect(host, port, NULL, 0, 5000, ssl ? TCP_SSL : 0, NULL);
  if(sb.sb_tcp == NULL)
    return HTTPRES_LOST_CONNECTION;
  i =
    sprintf(sb.sb_buf,
	    "GET %s HTTP/1.0\r\nUser-Agent: %s\r\nHost: %s\r\nReferer: %.*s\r\n",
	    path, AGENT, host, (int)(path - url + 1), url);
  if (http->date[0])
    i += sprintf(sb.sb_buf + i, "If-Modified-Since: %s\r\n", http->date);
  i += sprintf(sb.sb_buf + i, "\r\n");


  RTMPSockBuf_Send(&sb, sb.sb_buf, i);

  tcp_set_read_timeout(sb.sb_tcp, 5000);

  sb.sb_size = 0;
  sb.sb_timedout = FALSE;
  if (RTMPSockBuf_Fill(&sb) < 1)
    {
      ret = HTTPRES_LOST_CONNECTION;
      goto leave;
    }
  if (strncmp(sb.sb_buf, "HTTP/1", 6))
    {
      ret = HTTPRES_BAD_REQUEST;
      goto leave;
    }

  p1 = strchr(sb.sb_buf, ' ');
  rc = atoi(p1 + 1);
  http->status = rc;

  if (rc >= 300)
    {
      if (rc == 304)
	{
	  ret = HTTPRES_OK_NOT_MODIFIED;
	  goto leave;
	}
      else if (rc == 404)
	ret = HTTPRES_NOT_FOUND;
      else if (rc >= 500)
	ret = HTTPRES_SERVER_ERROR;
      else if (rc >= 400)
	ret = HTTPRES_BAD_REQUEST;
      else
	ret = HTTPRES_REDIRECTED;
    }

  p1 = memchr(sb.sb_buf, '\n', sb.sb_size);
  if (!p1)
    {
      ret = HTTPRES_BAD_REQUEST;
      goto leave;
    }
  sb.sb_start = p1 + 1;
  sb.sb_size -= sb.sb_start - sb.sb_buf;

  while ((p2 = memchr(sb.sb_start, '\r', sb.sb_size)))
    {
      if (*sb.sb_start == '\r')
	{
	  sb.sb_start += 2;
	  sb.sb_size -= 2;
	  break;
	}
      else
	if (!strncasecmp
	    (sb.sb_start, "Content-Length: ", sizeof("Content-Length: ") - 1))
	{
	  flen = atoi(sb.sb_start + sizeof("Content-Length: ") - 1);
	}
      else
	if (!strncasecmp
	    (sb.sb_start, "Last-Modified: ", sizeof("Last-Modified: ") - 1))
	{
	  *p2 = '\0';
	  strcpy(http->date, sb.sb_start + sizeof("Last-Modified: ") - 1);
	}
      p2 += 2;
      sb.sb_size -= p2 - sb.sb_start;
      sb.sb_start = p2;
      if (sb.sb_size < 1)
	{
	  if (RTMPSockBuf_Fill(&sb) < 1)
	    {
	      ret = HTTPRES_LOST_CONNECTION;
	      goto leave;
	    }
	}
    }

  len_known = flen > 0;
  while ((!len_known || flen > 0) &&
	 (sb.sb_size > 0 || RTMPSockBuf_Fill(&sb) > 0))
    {
      cb(sb.sb_start, 1, sb.sb_size, http->data);
      if (len_known)
	flen -= sb.sb_size;
      http->size += sb.sb_size;
      sb.sb_size = 0;
    }

  if (flen > 0)
    ret = HTTPRES_LOST_CONNECTION;

leave:
  RTMPSockBuf_Close(&sb);
  return ret;
}

#ifdef CRYPTO

#define CHUNK	16384

struct info
{
  z_stream *zs;
  HMAC_CTX ctx;
  int first;
  int zlib;
  int size;
};

static size_t
swfcrunch(void *ptr, size_t size, size_t nmemb, void *stream)
{
  struct info *i = stream;
  char *p = ptr;
  size_t len = size * nmemb;

  if (i->first)
    {
      i->first = 0;
      /* compressed? */
      if (!strncmp(p, "CWS", 3))
	{
	  *p = 'F';
	  i->zlib = 1;
	}
      HMAC_crunch(i->ctx, (unsigned char *)p, 8);
      p += 8;
      len -= 8;
      i->size = 8;
    }

  if (i->zlib)
    {
      unsigned char out[CHUNK];
      i->zs->next_in = (unsigned char *)p;
      i->zs->avail_in = len;
      do
	{
	  i->zs->avail_out = CHUNK;
	  i->zs->next_out = out;
	  inflate(i->zs, Z_NO_FLUSH);
	  len = CHUNK - i->zs->avail_out;
	  i->size += len;
	  HMAC_crunch(i->ctx, out, len);
	}
      while (i->zs->avail_out == 0);
    }
  else
    {
      i->size += len;
      HMAC_crunch(i->ctx, (unsigned char *)p, len);
    }
  return size * nmemb;
}

static int tzoff;
static int tzchecked;

#define	JAN02_1980	318340800

static const char *monthtab[12] = { "Jan", "Feb", "Mar",
  "Apr", "May", "Jun",
  "Jul", "Aug", "Sep",
  "Oct", "Nov", "Dec"
};
static const char *days[] =
  { "Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat" };

/* Parse an HTTP datestamp into Unix time */
static time_t
make_unix_time(char *s)
{
  struct tm time;
  int i, ysub = 1900, fmt = 0;
  char *month;
  char *n;
  time_t res;

  if (s[3] != ' ')
    {
      fmt = 1;
      if (s[3] != ',')
	ysub = 0;
    }
  for (n = s; *n; ++n)
    if (*n == '-' || *n == ':')
      *n = ' ';

  time.tm_mon = 0;
  n = strchr(s, ' ');
  if (fmt)
    {
      /* Day, DD-MMM-YYYY HH:MM:SS GMT */
      time.tm_mday = strtol(n + 1, &n, 0);
      month = n + 1;
      n = strchr(month, ' ');
      time.tm_year = strtol(n + 1, &n, 0);
      time.tm_hour = strtol(n + 1, &n, 0);
      time.tm_min = strtol(n + 1, &n, 0);
      time.tm_sec = strtol(n + 1, NULL, 0);
    }
  else
    {
      /* Unix ctime() format. Does not conform to HTTP spec. */
      /* Day MMM DD HH:MM:SS YYYY */
      month = n + 1;
      n = strchr(month, ' ');
      while (isspace(*n))
	n++;
      time.tm_mday = strtol(n, &n, 0);
      time.tm_hour = strtol(n + 1, &n, 0);
      time.tm_min = strtol(n + 1, &n, 0);
      time.tm_sec = strtol(n + 1, &n, 0);
      time.tm_year = strtol(n + 1, NULL, 0);
    }
  if (time.tm_year > 100)
    time.tm_year -= ysub;

  for (i = 0; i < 12; i++)
    if (!strncasecmp(month, monthtab[i], 3))
      {
	time.tm_mon = i;
	break;
      }
  time.tm_isdst = 0;		/* daylight saving is never in effect in GMT */

  /* this is normally the value of extern int timezone, but some
   * braindead C libraries don't provide it.
   */
  if (!tzchecked)
    {
      struct tm *tc;
      time_t then = JAN02_1980;
      tc = localtime(&then);
      tzoff = (12 - tc->tm_hour) * 3600 + tc->tm_min * 60 + tc->tm_sec;
      tzchecked = 1;
    }
  res = mktime(&time);
  /* Unfortunately, mktime() assumes the input is in local time,
   * not GMT, so we have to correct it here.
   */
  if (res != -1)
    res += tzoff;
  return res;
}

/* Convert a Unix time to a network time string
 * Weekday, DD-MMM-YYYY HH:MM:SS GMT
 */
static void
strtime(time_t * t, char *s)
{
  struct tm *tm;

  tm = gmtime((time_t *) t);
  sprintf(s, "%s, %02d %s %d %02d:%02d:%02d GMT",
	  days[tm->tm_wday], tm->tm_mday, monthtab[tm->tm_mon],
	  tm->tm_year + 1900, tm->tm_hour, tm->tm_min, tm->tm_sec);
}

#define HEX2BIN(a)      (((a)&0x40)?((a)&0xf)+9:((a)&0xf))

int
RTMP_HashSWF(const char *url, unsigned int *size, unsigned char *hash,
	     int age)
{
  FILE *f = NULL;
  char *path, date[64], cctim[64];
  long pos = 0;
  time_t ctim = -1, cnow;
  int i, got = 0, ret = 0;
  unsigned int hlen;
  struct info in = { 0 };
  struct HTTP_ctx http = { 0 };
  HTTPResult httpres;
  z_stream zs = { 0 };
  AVal home, hpre;

  date[0] = '\0';
#ifdef _WIN32
#ifdef XBMC4XBOX
  hpre.av_val = "Q:";
  hpre.av_len = 2;
  home.av_val = "\\UserData";
#else
  hpre.av_val = getenv("HOMEDRIVE");
  hpre.av_len = strlen(hpre.av_val);
  home.av_val = getenv("HOMEPATH");
#endif
#define DIRSEP	"\\"

#else /* !_WIN32 */
  hpre.av_val = "";
  hpre.av_len = 0;
  home.av_val = getenv("HOME");
#define DIRSEP	"/"
#endif
  if (!home.av_val)
    home.av_val = ".";
  home.av_len = strlen(home.av_val);

  /* SWF hash info is cached in a fixed-format file.
   * url: <url of SWF file>
   * ctim: HTTP datestamp of when we last checked it.
   * date: HTTP datestamp of the SWF's last modification.
   * size: SWF size in hex
   * hash: SWF hash in hex
   *
   * These fields must be present in this order. All fields
   * besides URL are fixed size.
   */
  path = malloc(hpre.av_len + home.av_len + sizeof(DIRSEP ".swfinfo"));
  sprintf(path, "%s%s" DIRSEP ".swfinfo", hpre.av_val, home.av_val);

  f = fopen(path, "r+");
  while (f)
    {
      char buf[4096], *file, *p;

      file = strchr(url, '/');
      if (!file)
	break;
      file += 2;
      file = strchr(file, '/');
      if (!file)
	break;
      file++;
      hlen = file - url;
      p = strrchr(file, '/');
      if (p)
	file = p;
      else
	file--;

      while (fgets(buf, sizeof(buf), f))
	{
	  char *r1;

	  got = 0;

	  if (strncmp(buf, "url: ", 5))
	    continue;
	  if (strncmp(buf + 5, url, hlen))
	    continue;
	  r1 = strrchr(buf, '/');
	  i = strlen(r1);
	  r1[--i] = '\0';
	  if (strncmp(r1, file, i))
	    continue;
	  pos = ftell(f);
	  while (got < 4 && fgets(buf, sizeof(buf), f))
	    {
	      if (!strncmp(buf, "size: ", 6))
		{
		  *size = strtol(buf + 6, NULL, 16);
		  got++;
		}
	      else if (!strncmp(buf, "hash: ", 6))
		{
		  unsigned char *ptr = hash, *in = (unsigned char *)buf + 6;
		  int l = strlen((char *)in) - 1;
		  for (i = 0; i < l; i += 2)
		    *ptr++ = (HEX2BIN(in[i]) << 4) | HEX2BIN(in[i + 1]);
		  got++;
		}
	      else if (!strncmp(buf, "date: ", 6))
		{
		  buf[strlen(buf) - 1] = '\0';
		  strncpy(date, buf + 6, sizeof(date));
		  got++;
		}
	      else if (!strncmp(buf, "ctim: ", 6))
		{
		  buf[strlen(buf) - 1] = '\0';
		  ctim = make_unix_time(buf + 6);
		  got++;
		}
	      else if (!strncmp(buf, "url: ", 5))
		break;
	    }
	  break;
	}
      break;
    }

  cnow = time(NULL);
  /* If we got a cache time, see if it's young enough to use directly */
  if (age && ctim > 0)
    {
      ctim = cnow - ctim;
      ctim /= 3600 * 24;	/* seconds to days */
      if (ctim < age)		/* ok, it's new enough */
	goto out;
    }

  in.first = 1;
  HMAC_setup(in.ctx, "Genuine Adobe Flash Player 001", 30);
  inflateInit(&zs);
  in.zs = &zs;

  http.date = date;
  http.data = &in;

  httpres = HTTP_get(&http, url, swfcrunch);

  inflateEnd(&zs);

  if (httpres != HTTPRES_OK && httpres != HTTPRES_OK_NOT_MODIFIED)
    {
      ret = -1;
      if (httpres == HTTPRES_LOST_CONNECTION)
	RTMP_Log(RTMP_LOGERROR, "%s: connection lost while downloading swfurl %s",
	    __FUNCTION__, url);
      else if (httpres == HTTPRES_NOT_FOUND)
	RTMP_Log(RTMP_LOGERROR, "%s: swfurl %s not found", __FUNCTION__, url);
      else
	RTMP_Log(RTMP_LOGERROR, "%s: couldn't contact swfurl %s (HTTP error %d)",
	    __FUNCTION__, url, http.status);
    }
  else
    {
      if (got && pos)
	fseek(f, pos, SEEK_SET);
      else
	{
	  char *q;
	  if (!f)
	    f = fopen(path, "w");
	  if (!f)
	    {
	      int err = errno;
	      RTMP_Log(RTMP_LOGERROR,
		  "%s: couldn't open %s for writing, errno %d (%s)",
		  __FUNCTION__, path, err, strerror(err));
	      ret = -1;
	      goto out;
	    }
	  fseek(f, 0, SEEK_END);
	  q = strchr(url, '?');
	  if (q)
	    i = q - url;
	  else
	    i = strlen(url);

	  fprintf(f, "url: %.*s\n", i, url);
	}
      strtime(&cnow, cctim);
      fprintf(f, "ctim: %s\n", cctim);

      if (!in.first)
	{
	  HMAC_finish(in.ctx, hash, hlen);
	  *size = in.size;

	  fprintf(f, "date: %s\n", date);
	  fprintf(f, "size: %08x\n", in.size);
	  fprintf(f, "hash: ");
	  for (i = 0; i < SHA256_DIGEST_LENGTH; i++)
	    fprintf(f, "%02x", hash[i]);
	  fprintf(f, "\n");
	}
    }
  HMAC_close(in.ctx);
out:
  free(path);
  if (f)
    fclose(f);
  return ret;
}
#else
int
RTMP_HashSWF(const char *url, unsigned int *size, unsigned char *hash,
	     int age)
{
  return -1;
}
#endif
